### ISSUE TITLE
[multistage] engine cleanup for operator on intermediate stage

### ISSUE THREAD

**Issue by walterddr** [apache/pinot#9026]
Summary
==
this issue Keeps track on the follow up / TODO list for operator implementation on multi-stage engine

Details
==
we tested basic mix-and-match operator SQLs but did not cover all combinations of possible SQL queries. This issue should focus on ensuring more SQL features are tested properly. 

Bug fixes:
- [x] timeout and other early termination doesn't polite GRPC mailbox channels (https://github.com/apache/pinot/pull/9116)
- [x] deal with large data block causes GRPC mailbox out-of-buffer exception (https://github.com/apache/pinot/pull/9116)
- [x] exception block population for timeouts (see: https://github.com/apache/pinot/pull/9064)
- [x] exception block population for GRPC mailbox failure. (see https://github.com/apache/pinot/pull/9143)
- [x] prevent broken mailbox channel (currently all mailbox reuse server-server channel, so one critical failure could cause all mailbox message lost) (see follow up on https://github.com/apache/pinot/pull/9143)
- [x] validate exchange reduction (see: https://github.com/apache/pinot/pull/9100)
- [x] reconnect after mailbox channel is broken (even after #9143 it can still have issues. check connections and reconnected mailbox channels and ensure proper channel closure)
  - [x] checked [grpc doc](https://grpc.github.io/grpc-java/javadoc/io/grpc/ConnectivityState.html), it should be automatically reconnect. the situation we ran into previously might've been a permanently stucked in `CONNECTING<->TRANSIENT_FAILURE` loop. will revisit if this actual occurs on production
- [x] early termination logic should trigger QueryDispatcher top-down cancellation on requestID (continue on https://github.com/apache/pinot/issues/9241)

**Comment by walterddr** [comment]
closing this one as all tickets are addressed (except for cancellation, which will continue in a separate issue since we already support this in normal broker request handler)
